core/TypePtr.h

```diff
@@ -4,25 +4,148 @@
 #include <memory>
 
 namespace sorbet::core {
-// using TypePtr = std::shared_ptr<Type>;
 class Type;
-class TypePtr {
-    std::shared_ptr<Type> store;
-    TypePtr(std::shared_ptr<Type> &&store);
+class TypePtr final {
+public:
+    // We store tagged pointers as 64-bit values.
+    using tagged_storage = u8;
+
+    enum class Tag {
+        ClassType = 1,
+        LambdaParam,
+        SelfTypeParam,
+        AliasType,
+        SelfType,
+        LiteralType,
+        TypeVar,
+        OrType,
+        AndType,
+        ShapeType,
+        TupleType,
+        AppliedType,
+        MetaType,
+        BlamedUntyped,
+        UnresolvedClassType,
+        UnresolvedAppliedType,
+    };
+
+    // A mapping from type to its corresponding tag.
+    template <typename T> struct TypeToTag;
+
+private:
+    std::atomic<u4> *counter;
+    tagged_storage store;
+
+    static constexpr tagged_storage TAG_MASK = 0xffff000000000007;
+
+    static constexpr tagged_storage PTR_MASK = ~TAG_MASK;
+
+    static tagged_storage tagPtr(Tag tag, void *expr) {
+        auto val = static_cast<tagged_storage>(tag);
+        if (val >= 8) {
+            // Store the tag in the upper 16 bits of the pointer, as it won't fit in the lower three bits.
+            val <<= 48;
+        }
+
+        auto maskedPtr = reinterpret_cast<tagged_storage>(expr) & PTR_MASK;
+
+        return maskedPtr | val;
+    }
+
+    explicit TypePtr(Tag tag, std::atomic<u4> *counter, void *expr) : counter(counter), store(tagPtr(tag, expr)) {
+        if (counter != nullptr) {
+            counter->fetch_add(1);
+        }
+    }
+    static void deleteTagged(Tag tag, void *ptr) noexcept;
+
+    // A version of release that doesn't mask the tag bits
+    tagged_storage releaseTagged() noexcept {
+        auto saved = store;
+        store = 0;
+        return saved;
+    }
+
+    std::atomic<u4> *releaseCounter() noexcept {
+        auto saved = counter;
+        counter = nullptr;
+        return saved;
+    }
+
+    void handleDelete() noexcept {
+        if (counter != nullptr) {
+            // fetch_sub returns value prior to subtract
+            const u4 counterVal = counter->fetch_sub(1) - 1;
```

Comment:
Did you decide to use `fetch_sub` instead of decrement because counter is a pointer?

Reply:
Yeah, I can't `--` a pointer without awkward dereffing. This seemed nicely explicit.